src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserHelperClasses.cs

```diff
@@ -300,103 +300,40 @@ internal void Clear()
 
     internal static class SqlEnvChangePool
     {
-        private static int _count;
-        private static int _capacity;
-        private static SqlEnvChange[] _items;
-#if DEBUG
-        private static string[] _stacks;
-#endif 
-
-        static SqlEnvChangePool()
+        private protected struct SqlEnvChangeWrapper
         {
-            _capacity = 10;
-            _items = new SqlEnvChange[_capacity];
-            _count = 1;
-#if DEBUG
-            _stacks = new string[_capacity];
-#endif 
+            public SqlEnvChange Item;
         }
 
+        private static readonly SqlEnvChangeWrapper[] _items = new SqlEnvChangeWrapper[Environment.ProcessorCount * 2];
+
         internal static SqlEnvChange Allocate()
         {
-            SqlEnvChange retval = null;
-            lock (_items)
+            SqlEnvChangeWrapper[] items = _items;
```

Comment:
Why make this copy? The array reference is readonly.

Reply:
Jit prefers local copies to class references for locality. not strictly needed, just a little better on the codegen.